_posts/2025-04-24-Redline-Testing-Comes-to-OpenSearch-Benchmark.md

```diff
@@ -0,0 +1,163 @@
+
+
+## The challenge of guessing thresholds
+
+OpenSearch Benchmark recently introduced support for ramping up clients and throughput during a benchmark run. However, users still needed to perform the following actions when benchmarking:
+
+- Estimate what amount of load would break the cluster.  
+- Manually watch for failures in logs or dashboards.  
+- Rerun benchmarks with different parameters.  
+
+These time-consuming tasks made it difficult to pinpoint the maximum sustainable throughput.
+
+## What we built: Redline testing
+
+Redline testing automates the following:
+
+- A **Feedback Actor** to monitor request failures in real time  
+- A **self-adjusting load mechanism** to pause or unpause clients based on observed behavior  
+
+This enables OpenSearch Benchmark to perform the following actions:
+
+- Ramp up active clients  
+- Detect when failures begin  
+- Scale back automatically, wait for recovery, and resume testing  
+
+The result: OpenSearch Benchmark can now determine your cluster's redline in a single test run.
+
+## How it works
+
+The following diagram provides a high-level overview of the actor-based execution flow in OpenSearch Benchmark. It illustrates how benchmark tasks are allocated and executed across multiple worker actors, each of which manages a group of clients responsible for sending requests to a target OpenSearch cluster, using the following components:
+
+- **BenchmarkActor**: Initiates the benchmarking process.  
+- **WorkerCoordinatorActor**: Manages worker lifecycle and task distribution using an allocation matrix from the **Allocator**.  
+- **Workers (Worker1 through WorkerN)**: Executes tasks by managing clients via the **AsyncIoAdapter**.  
+- **Clients (Client1 through ClientN)**: Uses the `AsyncExecutor` class to perform operations against the target host in parallel.  
+
+![Flowchart for OpenSearch Benchmark's actor system](/assets/media/blog-images/2025-04-24-Redline-Testing-Comes-to-OpenSearch-Benchmark/OpenSearch Benchmark-system-architecture.jpg){: .img-fluid}
+
+OpenSearch Benchmark uses the **Actor Model**, which structures concurrent, distributed systems around isolated, message-passing components.
+
+Each actor performs the following actions:
+
+- Maintains private, isolated state  
+- Processes messages sequentially  
+- Can create other actors  
+- Communicates asynchronously  
+
+By default, OpenSearch Benchmark spawns `n` worker actors (one per CPU core), distributing clients evenly across them. For example, with 10 workers and 20 clients, each manages 2 clients.
+ 
+
+### Timed mode and redline logic
+
+OpenSearch Benchmark supports the following two modes for redline testing:
+
+- **Iteration mode**: Runs tasks for a fixed number of iterations  
+- **Timed mode**: Runs tasks for a fixed duration  
+
+Redline testing uses timed mode to perform the following actions:
```

Comment:
Redline testing operates only in the timed mode.  It performs the following actions: